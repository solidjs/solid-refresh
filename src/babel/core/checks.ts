import * as t from '@babel/types';

// This is just a Pascal heuristic
// we only assume a function is a component
// if the first character is in uppercase
export function isComponentishName(name: string) {
  return name[0] >= 'A' && name[0] <= 'Z';
}

export function getImportSpecifierName(specifier: t.ImportSpecifier): string {
  if (t.isIdentifier(specifier.imported)) {
    return specifier.imported.name;
  }
  return specifier.imported.value;
}
