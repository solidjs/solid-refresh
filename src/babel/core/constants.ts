import type { ImportDefinition, ImportIdentifierSpecifier } from './types';

// Source of solid-refresh (for import)
const SOLID_REFRESH_MODULE = 'solid-refresh';

// Exported names from solid-refresh that will be imported
export const IMPORT_REGISTRY: ImportDefinition = {
  kind: 'named',
  name: '$$registry',
  source: SOLID_REFRESH_MODULE,
};

export const IMPORT_REFRESH: ImportDefinition = {
  kind: 'named',
  name: '$$refresh',
  source: SOLID_REFRESH_MODULE,
};

export const IMPORT_COMPONENT: ImportDefinition = {
  kind: 'named',
  name: '$$component',
  source: SOLID_REFRESH_MODULE,
};

export const IMPORT_CONTEXT: ImportDefinition = {
  kind: 'named',
  name: '$$context',
  source: SOLID_REFRESH_MODULE,
};

export const IMPORT_DECLINE: ImportDefinition = {
  kind: 'named',
  name: '$$decline',
  source: SOLID_REFRESH_MODULE,
};

export const IMPORT_SPECIFIERS: ImportIdentifierSpecifier[] = [
  {
    type: 'render',
    definition: { name: 'render', kind: 'named', source: 'solid-js/web' },
  },
  {
    type: 'render',
    definition: { name: 'hydrate', kind: 'named', source: 'solid-js/web' },
  },
  {
    type: 'createContext',
    definition: {
      name: 'createContext',
      kind: 'named',
      source: 'solid-js',
    },
  },
  {
    type: 'createContext',
    definition: {
      name: 'createContext',
      kind: 'named',
      source: 'solid-js/web',
    },
  },
];
