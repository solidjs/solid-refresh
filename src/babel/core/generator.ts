import _generator from '@babel/generator';
import type * as t from '@babel/types';

// https://github.com/babel/babel/issues/15269
let generator: typeof _generator;
if (typeof _generator !== 'function') {
  generator = (_generator as any).default;
} else {
  generator = _generator;
}

export function generateCode(node: t.Node): string {
  return generator(node).code;
}
