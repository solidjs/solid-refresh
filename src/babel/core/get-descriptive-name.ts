import type * as babel from '@babel/core';

export function getDescriptiveName(
  path: babel.NodePath,
  defaultName: string,
): string {
  let current: babel.NodePath | null = path;
  while (current) {
    switch (current.node.type) {
      case 'FunctionDeclaration':
      case 'FunctionExpression': {
        if (current.node.id) {
          return current.node.id.name;
        }
        break;
      }
      case 'VariableDeclarator': {
        if (current.node.id.type === 'Identifier') {
          return current.node.id.name;
        }
        break;
      }
      case 'ClassPrivateMethod':
      case 'ClassMethod':
      case 'ObjectMethod': {
        switch (current.node.key.type) {
          case 'Identifier':
            return current.node.key.name;
          case 'PrivateName':
            return current.node.key.id.name;
          default:
            break;
        }
        break;
      }
      default:
        break;
    }
    current = current.parentPath;
  }
  return defaultName;
}
