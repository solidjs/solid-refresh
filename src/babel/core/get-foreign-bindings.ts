import * as t from '@babel/types';
import type * as babel from '@babel/core';

function isForeignBinding(
  source: babel.NodePath,
  current: babel.NodePath,
  name: string,
): boolean {
  if (source === current) {
    return true;
  }
  if (current.scope.hasOwnBinding(name)) {
    return false;
  }
  if (current.parentPath) {
    return isForeignBinding(source, current.parentPath, name);
  }
  return true;
}

function isInTypescript(path: babel.NodePath): boolean {
  let parent = path.parentPath;
  while (parent) {
    if (t.isTypeScript(parent.node) && !t.isExpression(parent.node)) {
      return true;
    }
    parent = parent.parentPath;
  }
  return false;
}

export function getForeignBindings(path: babel.NodePath): t.Identifier[] {
  const identifiers = new Set<string>();
  path.traverse({
    ReferencedIdentifier(p) {
      // Check identifiers that aren't in a TS expression
      if (!isInTypescript(p) && isForeignBinding(path, p, p.node.name)) {
        if (p.isIdentifier() || p.parentPath.isJSXMemberExpression()) {
          identifiers.add(p.node.name);
        }
      }
    },
  });
  const collected = [];
  for (const identifier of identifiers) {
    collected.push(t.identifier(identifier));
  }
  return collected;
}
