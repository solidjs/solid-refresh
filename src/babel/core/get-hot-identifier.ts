import type { StateContext } from './types';
import * as t from '@babel/types';

export function getHotIdentifier(state: StateContext): t.MemberExpression {
  switch (state.opts.bundler) {
    // vite/esm uses `import.meta.hot`
    case 'esm':
    case 'vite':
      return t.memberExpression(
        t.memberExpression(t.identifier('import'), t.identifier('meta')),
        t.identifier('hot'),
      );
    // webpack 5 uses `import.meta.webpackHot`
    // rspack does as well
    case 'webpack5':
    case 'rspack-esm':
      return t.memberExpression(
        t.memberExpression(t.identifier('import'), t.identifier('meta')),
        t.identifier('webpackHot'),
      );
    default:
      // `module.hot` is the default.
      return t.memberExpression(t.identifier('module'), t.identifier('hot'));
  }
}
