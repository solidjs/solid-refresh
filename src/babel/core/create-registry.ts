import type * as babel from '@babel/core';
import * as t from '@babel/types';
import { IMPORT_REFRESH, IMPORT_REGISTRY } from './constants';
import { getHotIdentifier } from './get-hot-identifier';
import { getImportIdentifier } from './get-import-identifier';
import { getRootStatementPath } from './get-root-statement-path';
import { generateViteHMRRequirement } from './get-vite-hmr-requirement';
import type { StateContext } from './types';

const REGISTRY = 'REGISTRY';

export function createRegistry(
  state: StateContext,
  path: babel.NodePath,
): t.Identifier {
  const current = state.imports.get(REGISTRY);
  if (current) {
    return current;
  }
  const root = getRootStatementPath(path);
  const identifier = path.scope.generateUidIdentifier(REGISTRY);

  root.scope.registerDeclaration(
    root.insertBefore(
      t.variableDeclaration('const', [
        t.variableDeclarator(
          identifier,
          t.callExpression(
            getImportIdentifier(state, path, IMPORT_REGISTRY),
            [],
          ),
        ),
      ]),
    )[0],
  );
  const pathToHot = getHotIdentifier(state);
  const statements: t.Statement[] = [
    t.expressionStatement(
      t.callExpression(getImportIdentifier(state, path, IMPORT_REFRESH), [
        t.stringLiteral(state.bundler),
        pathToHot,
        identifier,
      ]),
    ),
  ];

  generateViteHMRRequirement(state, statements, pathToHot);

  (
    path.scope.getProgramParent().path as babel.NodePath<t.Program>
  ).pushContainer('body', [
    t.ifStatement(pathToHot, t.blockStatement(statements)),
  ]);
  state.imports.set(REGISTRY, identifier);
  return identifier;
}
