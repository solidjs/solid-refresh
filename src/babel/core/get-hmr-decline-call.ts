import type * as babel from '@babel/core';
import * as t from '@babel/types';
import type { StateContext } from './types';
import { getHotIdentifier } from './get-hot-identifier';
import { getImportIdentifier } from './get-import-identifier';
import { IMPORT_DECLINE } from './constants';
import { generateViteHMRRequirement } from './get-vite-hmr-requirement';

export function getHMRDeclineCall(state: StateContext, path: babel.NodePath) {
  const pathToHot = getHotIdentifier(state);
  const statements = [
    t.expressionStatement(
      t.callExpression(getImportIdentifier(state, path, IMPORT_DECLINE), [
        t.stringLiteral(state.opts.bundler ?? 'standard'),
        pathToHot,
      ]),
    ),
  ];

  generateViteHMRRequirement(state, statements, pathToHot);

  return t.ifStatement(pathToHot, t.blockStatement(statements));
}
