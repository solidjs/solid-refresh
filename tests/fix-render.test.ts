import * as babel from "@babel/core";
import { describe, it, expect } from "vitest";
import plugin from "../src/babel";

async function transform(code: string) {
  const result = await babel.transformAsync(code, {
    plugins: [[plugin, { bundler: "vite" }]],
    parserOpts: {
      plugins: ["jsx", "typescript"]
    }
  });

  if (result && result.code) {
    return result.code;
  }
  throw new Error("Missing code");
}

describe("fix render", () => {
  describe("import specifiers", () => {
    it("should work with ImportSpecifier + Identifier", async () => {
      expect(
        await transform(`
        import { render } from 'solid-js/web';
  
        render(() => <App />, root);
      `)
      ).toMatchSnapshot();
    });
    it("should work with ImportSpecifier + aliased Identifier", async () => {
      expect(
        await transform(`
        import { render as Render } from 'solid-js/web';
  
        Render(() => <App />, root);
      `)
      ).toMatchSnapshot();
    });
    it("should work with ImportSpecifier + aliased Identifier from StringLiteral", async () => {
      expect(
        await transform(`
        import { 'render' as Render } from 'solid-js/web';
  
        Render(() => <App />, root);
      `)
      ).toMatchSnapshot();
    });
    it("should work with ImportNamespaceSpecifier", async () => {
      expect(
        await transform(`
        import * as solidWeb from 'solid-js/web';
  
        solidWeb.render(() => <App />, root);
      `)
      ).toMatchSnapshot();
    });
  });
  describe("top-level statements", async () => {
    it("should work with IfStatement", async () => {
      expect(
        await transform(`
        import { render } from 'solid-js/web';

        if (root) {
          render(() => <App />, root);
        }
      `)
      ).toMatchSnapshot();
    });
  });
});
