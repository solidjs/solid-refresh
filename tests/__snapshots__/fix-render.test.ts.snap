// Vitest Snapshot v1

exports[`fix render > import specifiers > should work with ImportNamespaceSpecifier 1`] = `
"import { DEV as _DEV } from \\"solid-refresh\\";
import * as solidWeb from 'solid-js/web';

const _cleanup = solidWeb.render(() => <App />, root);

if (import.meta.hot) import.meta.hot.dispose(_cleanup);

if (!(_DEV && Object.keys(_DEV).length)) {
  console.warn(\\"To use solid-refresh, you need to use the dev build of SolidJS. Make sure your build system supports package.json conditional exports and has the 'development' condition turned on.\\");
  if (import.meta.hot) import.meta.hot.decline();
}"
`;

exports[`fix render > import specifiers > should work with ImportSpecifier + Identifier 1`] = `
"import { DEV as _DEV } from \\"solid-refresh\\";
import { render } from 'solid-js/web';

const _cleanup = render(() => <App />, root);

if (import.meta.hot) import.meta.hot.dispose(_cleanup);

if (!(_DEV && Object.keys(_DEV).length)) {
  console.warn(\\"To use solid-refresh, you need to use the dev build of SolidJS. Make sure your build system supports package.json conditional exports and has the 'development' condition turned on.\\");
  if (import.meta.hot) import.meta.hot.decline();
}"
`;

exports[`fix render > import specifiers > should work with ImportSpecifier + aliased Identifier 1`] = `
"import { DEV as _DEV } from \\"solid-refresh\\";
import { render as Render } from 'solid-js/web';

const _cleanup = Render(() => <App />, root);

if (import.meta.hot) import.meta.hot.dispose(_cleanup);

if (!(_DEV && Object.keys(_DEV).length)) {
  console.warn(\\"To use solid-refresh, you need to use the dev build of SolidJS. Make sure your build system supports package.json conditional exports and has the 'development' condition turned on.\\");
  if (import.meta.hot) import.meta.hot.decline();
}"
`;

exports[`fix render > import specifiers > should work with ImportSpecifier + aliased Identifier from StringLiteral 1`] = `
"import { DEV as _DEV } from \\"solid-refresh\\";
import { 'render' as Render } from 'solid-js/web';

const _cleanup = Render(() => <App />, root);

if (import.meta.hot) import.meta.hot.dispose(_cleanup);

if (!(_DEV && Object.keys(_DEV).length)) {
  console.warn(\\"To use solid-refresh, you need to use the dev build of SolidJS. Make sure your build system supports package.json conditional exports and has the 'development' condition turned on.\\");
  if (import.meta.hot) import.meta.hot.decline();
}"
`;

exports[`fix render > top-level statements > should work with IfStatement 1`] = `
"import { DEV as _DEV } from \\"solid-refresh\\";
import { render } from 'solid-js/web';

if (root) {
  const _cleanup = render(() => <App />, root);

  if (import.meta.hot) import.meta.hot.dispose(_cleanup);
}

if (!(_DEV && Object.keys(_DEV).length)) {
  console.warn(\\"To use solid-refresh, you need to use the dev build of SolidJS. Make sure your build system supports package.json conditional exports and has the 'development' condition turned on.\\");
  if (import.meta.hot) import.meta.hot.decline();
}"
`;
